### PR TITLE
Introduce parquet file format for bulk filesystem

### PR DESCRIPTION
**Pull request by LesTR**
Closes https://github.com/O2-Czech-Republic/proxima-platform/issues/186

### CODE REVIEW

File: scheme/proto/src/test/resources/test-proto.conf [hunk 1]
```diff
@@ -4,6 +4,7 @@
     event: {
       attributes: {
         data: { scheme: "proto:cz.o2.proxima.scheme.proto.test.Scheme.Event" }
+        complex: {scheme: "proto:cz.o2.proxima.scheme.proto.test.Scheme.ValueSchemeMessage"}
```
Comment (dmvk) [inline scheme/proto/src/test/resources/test-proto.conf:7]: nit: inconsitent whitespaces (also in changes below)

File: direct/io-bulkfs/src/test/java/cz/o2/proxima/direct/bulk/AbstractFileFormatTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+@Slf4j
+public abstract class AbstractFileFormatTest {
```
Comment (dmvk) [inline direct/io-bulkfs/src/test/java/cz/o2/proxima/direct/bulk/AbstractFileFormatTest.java:50]: I love to see, that we finally have a common test suite for file formats 👍

File: direct/io-bulkfs/src/main/java/cz/o2/proxima/direct/bulk/FileFormatUtils.java [hunk 3]
```diff
@@ -97,6 +97,10 @@ static FileFormat getFileFormatFromName(String format, boolean gzip) {
     if ("json".equals(format)) {
       return FileFormat.json(gzip);
     }
+    if ("parquet".equals(format)) {
+      // just an shortcut for parquet file format
+      format = "cz.o2.proxima.direct.bulk.fs.parquet.ParquetFileFormat";
```
Comment (dmvk) [inline direct/io-bulkfs/src/main/java/cz/o2/proxima/direct/bulk/FileFormatUtils.java:102]:
```suggestion
      return new ParquetFileFormat();
```

Why don't you instantiate the format immediately?

File: direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/TestRecordConsumer.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.bulk.fs.parquet;
+
+import lombok.Getter;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.io.api.RecordConsumer;
+
+class TestRecordConsumer extends RecordConsumer {
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/TestRecordConsumer.java:22]: nice 👍

File: direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementWriteSupportTest.java [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+                    Repetition.REQUIRED,
+                    "test-schema",
+                    ParquetUtils.mapSchemaTypeToParquet(descriptor, attributeName)));
+    writer.setName(attributeName);
+    writer.setIndex(index);
+    writer.write(value);
+    assertEquals(expected, recordConsumer.getWritten());
+  }
+
+  private static class TestRecordConsumer extends RecordConsumer {
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementWriteSupportTest.java:374]: Either this or `cz.o2.proxima.direct.bulk.fs.parquet.TestRecordConsumer` should be removed

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  public static final String PARQUET_CONFIG_PAGE_SIZE_KEY_NAME = ParquetOutputFormat.PAGE_SIZE;
+
+  /**
+   * Config key used for attribute prefix - useful when attribute name collides with proxima fields
+   * defined as PARQUET_COLUMN_NAME_*
+   */
+  public static final String PARQUET_CONFIG_VALUES_PREFIX_KEY_NAME = "parquet.values.name.prefix";
+
+  /** Name for top level message in parquet schema. */
+  public static final String PARQUET_MESSAGE_NAME = "proxima-bulk";
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:57]: Maybe stream element?

```suggestion
  public static final String PARQUET_MESSAGE_NAME = "stream_element";
```

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        getParquetSchema());
+    return new ProximaParquetWriter(
+        path,
+        getParquetSchema(),
+        attributeNamesPrefix,
+        parquetCompressionCodec,
+        createWriterConfiguration());
+  }
+
+  Configuration createWriterConfiguration() {
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:130]: is it worth anything to cache & reuse the configuration instance?

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  /** Operation column name in parquet file. See {@link Operation} for details. */
+  static final String PARQUET_COLUMN_NAME_OPERATION = "operation";
+
+  /** Compression for parquet file */
+  CompressionCodecName parquetCompressionCodec;
+
+  @Nullable private transient MessageType parquetSchema;
+  private AttributeFamilyDescriptor familyDescriptor;
+  private String attributeNamesPrefix;
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:88]: Should all properties, that are filled in `setup` be marked as transient?

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+  @Override
+  public void setup(AttributeFamilyDescriptor family) {
+    familyDescriptor = family;
+    parquetCompressionCodec =
+        CompressionCodecName.fromConf(
+            Optional.ofNullable(family.getCfg().get(PARQUET_CONFIG_COMPRESSION_KEY_NAME))
+                .map(Object::toString)
+                .orElse(
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:97]: nit: orElseGet

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    return parquetSchema;
+  }
+
+  @Override
+  public String fileSuffix() {
+    return "parquet" + parquetCompressionCodec.getExtension();
+  }
+
+  /** Enum used for mapping proxima operation into (and from) parquet files. */
+  enum Operation {
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:159]: nit: enums should be defined under `static` variables section

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    UPSERT("u"),
+    DELETE("d"),
+    DELETE_WILDCARD("w");
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:162]: nit: "dw"?

**Comment by je-ik** [comment]
Would it be possible to add comments to review and send the complete review as single message? Would help a lot to reduce email spamming. :-)

File: direct/io-bulkfs/src/main/java/cz/o2/proxima/direct/bulk/FileFormatUtils.java (hunk 3 above)
Comment (LesTR) [reply direct/io-bulkfs/src/main/java/cz/o2/proxima/direct/bulk/FileFormatUtils.java:102]: We can't do that because it will create circular dependency between `proxima-direct-io-bulkfs` and `proxima-direct-io-bulkfs-parquet`

File: direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementWriteSupportTest.java (hunk 5 above)
Comment (LesTR) [reply direct/io-bulkfs-parquet/src/test/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementWriteSupportTest.java:374]: Thanks, I forgot this during refactoring.

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java (hunk 10 above)
Comment (LesTR) [reply direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:159]: nested enum types are implicitly static

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java (hunk 9 above)
Comment (LesTR) [reply direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:99]: That's is not needed because it is wrapped by `CompressionCodecName.fromConf()` which returns default settings (uncompressed)

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java (hunk 10 above)
Comment (dmvk) [reply direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ParquetFileFormat.java:167]: yes, I just wanted to move this block up in the file, where the static definition belongs :)

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ProximaParquetReader.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    final SeekableByteChannel channel = (SeekableByteChannel) path.read();
+    final Configuration configuration = new Configuration(false);
+    this.reader =
+        new ParquetReadBuilder(new BulkInputFile(channel), entity)
+            .withConf(configuration)
+            /**
+             * Currently we can not use push down filter for attributes See
+             * https://github.com/O2-Czech-Republic/proxima-platform/issues/196 for details
+             * .withFilter()
+             */
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/ProximaParquetReader.java:57]: nit: block comments are for javadoc only, use `// ...` inside methods instead

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementMaterializer.java [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+    final AttributeValueAccessor<Object, Object> valueAccessor = valueSerializer.getValueAccessor();
+    final SchemaTypeDescriptor<Object> attributeSchema = valueSerializer.getValueSchemaDescriptor();
+    Object attributeValue;
+    if (attributeSchema.isStructureType()) {
+      attributeValue = record.getOrDefault(storedAttributeName, new HashMap<>());
+      @SuppressWarnings("unchecked")
+      final StructureValue cast = StructureValue.of((Map<String, Object>) attributeValue);
+      return valueSerializer.serialize(valueAccessor.createFrom(cast));
+    } else {
+      attributeValue = record.getOrDefault(storedAttributeName, new Object());
```
Comment (dmvk) [inline direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementMaterializer.java:125]: can you please explain this default?

**Review by dmvk — APPROVED** [review]
LGTM 👍 This is a really great change set. I did not cover all the edge cases, but all the critical code path seems to be stressed to a great degree by the newly added test suite. Also we've already tested this on a live deployment.

Also comparing to prior versions, there is great simplification in the codebase and it become way easier to understand whats going on.

I've added few minor comments, but I think it's OK to merge this as is. 

Good job 🎉 👏

File: direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementMaterializer.java (hunk 13 above)
Comment (LesTR) [reply direct/io-bulkfs-parquet/src/main/java/cz/o2/proxima/direct/bulk/fs/parquet/StreamElementMaterializer.java:125]: Added note into code:
```
There is nothing written into parquet file when we are writing empty element. In that case we need to delegate creation of empty element to serializer. This is reason for that record.getOrDefault. 
```
Hopefully it is more clear now.
